crate_anon/preprocess/autoimport_db.py

```diff
@@ -0,0 +1,1167 @@
+        SQLAlchemy will complain that we re-assign columns to a table). Assumes
+        that the MetaData object WILL NOT CHANGE ACROSS CALLS.
+        """
+        if self._table is None:
+            self._table = Table(
+                self.tablename, self.metadata, *(self.columns or ())
+            )
+        return self._table
+
+    def drop_table(self) -> None:
+        """
+        Drop a database table, if it exists.
+        """
+        log.info(f"Dropping table (if it exists): {self.tablename}")
+        # See also crate_anon.anonymise.subset_db.drop_dst_table_if_exists().
+        t = self.table()  # doesn't need columns
+        t.drop(self.engine, checkfirst=True)
+        # No COMMIT required after DDL.
+        self.metadata.remove(t)  # otherwise we may struggle to re-create it
+        self.table_exists_in_database = False
+
+    def create_table(self) -> None:
+        """
+        Create a database table, if it doesn't already exist.
+        """
+        if self.table_exists_in_database:
+            log.info(f"Table already exists, not creating: {self.tablename}")
+            return
+        log.info(f"Creating table: {self.tablename}\n{self.colreport()}")
+        t = self.table()
+        try:
+            t.create(self.engine, checkfirst=True)
+        except InvalidRequestError:
+            log.warning(
+                "Table already exists, unexpectedly; not re-creating: "
+                f"{self.tablename}"
+            )
+        # No COMMIT required after DDL.
+
+
+class ColumnTypeDetector:
+    """
+    Class to inspect values from a spreadsheet column, and make a decision
+    about what kind of SQLAlchemy column should be used.
+    """
+
+    def __init__(
+        self,
+        colname: str,
+        values: Iterable[Any] = None,
+        default_type: TypeEngine = None,
+        verbose: bool = False,
+    ) -> None:
+        """
+        Args:
+            colname:
+                Future column name.
+            values:
+                Optional values to inspect immediately.
+            default_type:
+                Type to use if no non-NULL data whatsoever is seen.
+            verbose:
+                Report values on failure. WARNING: unsuitable, for this reason,
+                for production code.
+        """
+        if not colname:
+            raise ValueError("Missing column name")
+        if "\t" in colname:
+            raise ValueError(
+                f"Likely TSV file being read as CSV, because column name "
+                f"includes tabs: {colname!r}"
+            )
+
+        self.colname = colname
+        self.default_type = default_type
+        self.verbose = verbose
+
+        # None/NULL types:
+        self.seen_none = False
+
+        # Sanity check
+        self.seen_something_not_none = False
+
+        # Non-string types:
+        self.seen_bool = False
+        self.seen_int = False
+        self.seen_float = False
+        # self.min_numeric = math.inf
```

Comment:
Remove commented code if not used or include with command line parameter if it's helpful to have it? It's still in version control if we need it.